src/wallet/wallet2.cpp

```diff
@@ -1049,7 +1049,9 @@ void wallet2::process_new_transaction(const crypto::hash &txid, const cryptonote
   }
 
   uint64_t tx_money_spent_in_ins = 0;
-  boost::optional<uint32_t> subaddr_account;
+  // The line below is equivalent to "boost::optional<uint32_t> subaddr_account;", but avoids the GCC warning: ‘*((void*)& subaddr_account +4)’ may be used uninitialized in this function
+  // It's a GCC bug with boost::optional, see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=47679
+  auto subaddr_account ([]()->boost::optional<uint32_t> {return boost::none;}());
```

Comment:
Is that different from boost::optional<uint32_t> subaddr_account = boost::none; ?

Reply:
"boost::optional<uint32_t> subaddr_account = boost::none;" does not silence the warning.   
   
I think it's because it assigns a "normal" value but leaves the boost::optional specific part uninitialized (which is the one taken into account with a boolean check `if (subaddr_accoun)`). boost::optional variables default to false for this test, but GCC doesn't take it into account. This can only be explicitly initialized by a value returned by a function, thus the lambda. I'm not sure how to explain better, and I'm also not super familiar with it.

See https://stackoverflow.com/questions/21755206/how-to-get-around-gcc-void-b-4-may-be-used-uninitialized-in-this-funct

Reply:
In other words, setting subaddr_account = boost::none would work for testing `if (*subaddr_account)` but not `if (subaddr_account)`.

Reply:
OK. That page has a simpler workaround (boost::make_optional) fwiw.

Reply:
I did not want to use that other workaround because IIUC it lets the compiler define the type from a value only, and I am not sure how to cleanly make sure it picks `uint32_t`. While with the workaround in this patch, despite the "auto" you can explicitly tell the compiler in that same line that it must be an uint32_t.

Reply:
> despite the "auto" you can explicitly tell the compiler in that same line that it must be an uint32_t.

I personally prefer that (mainly for the sake of ease of read): `boost::optional<uint32_t> subaddr_account ([]()...);`

Reply:
`make_optional<uint32_t>(false, 0)` or `make_optional(false, uint32_t(0))`

Reply:
@vtnerd 

```boost::optional<uint32_t> subaddr_account;
  subaddr_account = boost::make_optional(false, uint32_t(0));```

Did not silence the message. (gcc 5.4.0).

Reply:
BTW, what version of boost is this? At some point in the recent past the `boost::optional` implementation changed too.

Reply:
I'm on 1.58.0